pkgs/os-specific/serenity/libc.nix

```diff
@@ -0,0 +1,105 @@
+{ lib
+, stdenv
+, callPackage
+, runCommand
+, pkgsBuildBuild
+, buildPackages
+}:
+
+let
+  inherit (callPackage ./common.nix { }) src version;
+
+  cpuArch = stdenv.targetPlatform.parsed.cpu.name;
+in
+stdenv.mkDerivation rec {
+  inherit src version;
+  pname = "serenitylibc";
+
+  postPatch = ''
+    substituteInPlace Toolchain/CMake/GNUToolchain.txt.in \
+      --replace "gcc-ar" "ar" \
+      --replace "gcc-nm" "nm" \
+      --replace "gcc-ranlib" "ranlib"
+    echo "" > Userland/Libraries/LibC/stubs.cpp
+    echo "" > Tests/LibELF/CMakeLists.txt
+  '';
+
+  depsBuildBuild =
+    [
+      buildPackages.stdenv.cc
+      pkgsBuildBuild.cmake
+      pkgsBuildBuild.ninja
+    ]
+    ++ lib.optional stdenv.buildPlatform.isLinux buildPackages.libxcrypt;
+
+  # The default cmake configure phase sets CMAKE_INSTALL_* flags that interfere
+  # with expected paths
+  dontUseCmakeConfigure = true;
+  configurePhase = ''
+    runHook preConfigure
+
+    mkdir -p Toolchain/Local
+    ln -s ${stdenv.cc} Toolchain/Local/${cpuArch}
+
+    cmake -S Meta/CMake/Superbuild -B Build/superbuild-${cpuArch} -GNinja \
+      -DCMAKE_C_COMPILER=gcc \
+      -DCMAKE_CXX_COMPILER=g++ \
+      -DSERENITY_ARCH=${cpuArch} \
+      -DSERENITY_TOOLCHAIN=GNU \
+      -DENABLE_TIME_ZONE_DATABASE_DOWNLOAD=OFF \
+      -DENABLE_UNICODE_DATABASE_DOWNLOAD=OFF \
+      -DENABLE_PCI_IDS_DOWNLOAD=OFF \
+      -DENABLE_USB_IDS_DOWNLOAD=OFF \
+      -DENABLE_PNP_IDS_DOWNLOAD=OFF \
+      -DENABLE_CACERT_DOWNLOAD=OFF
+
+    runHook postConfigure
+  '';
+
+  ninjaFlagsArray = [ "-CBuild/superbuild-${cpuArch}" ];
+  enableParallelBuilding = true;
+
+  # Build system uses --sysroot=Build/<arch>/Root to specify location of crt
+  # libraries but the flag is unsupported in nixpkgs's ld-wrapper. Use -B as
+  # a workaround
+  preBuild = ''
+    export NIX_CFLAGS_COMPILE+=" -B$(pwd)/Build/${cpuArch}/Root/usr/lib/"
+  '';
+  hardeningDisable = [ "fortify" ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/lib
+    cp Build/${cpuArch}/Root/usr/lib/{crt*,libc.*,libdl.*,libm.*,libpthread.*,libssp.*,libsystem.*,Loader.*} $out/lib
+    cp -r ${passthru.headers}/include $out/include
+
+    runHook postInstall
+  '';
+
+  passthru = {
+    headers = runCommand "serenity-headers" {}
+      ''
+        cd ${src}
+        FILES=$(find \
+                AK \
+                Kernel/API \
+                Kernel/Arch \
+                Userland/Libraries/LibC \
+                -name '*.h' -print)
+        for header in $FILES; do
+          target=$(echo "$header" | sed -e "s|Userland/Libraries/LibC||")
+          mkdir -p "$(dirname "$out/include/$target")"
+          cp "$header" "$out/include/$target"
+        done
+      '';
```

Comment:
I'd rather this lived in its own file, and was a proper package with a meta attribute and stuff.